drivers/hmis/app/models/hmis/file.rb

```diff
@@ -43,17 +43,23 @@ def data_source
   scope :confidential, -> { where(confidential: true) }
   scope :nonconfidential, -> { where(confidential: [false, nil]) }
 
-  scope :viewable_by, ->(user) do
+  # Passing client_ids can significantly improve performance, to avoid a subquery on the entire Client table.
+  # This may not be the perfect solution (combining scopes would be neater) but it works.
+  # It is similar to the strategy described in the Warehouse's EnrollmentArbiter.
+  scope :viewable_by, ->(user, client_ids: nil) do
     # NOTE: it's okay that confidential files are included in this scope even if the user
     # doesn't have permission to read the file. Users can see the existence of confidential
-    # files but they can't read them. Reference:
-    # https://github.com/open-path/Green-River/issues/5184
+    # files but they can't read them. Reference: https://github.com/open-path/Green-River/issues/5184
+
     client_scope = Hmis::Hud::Client.
       viewable_by(user).
       with_access(user, :can_view_any_nonconfidential_client_files, :can_view_any_confidential_client_files)
+    client_scope = client_scope.where(id: client_ids) if client_ids.present?
+
     enrollment_scope = Hmis::Hud::Enrollment.
       viewable_by(user).
       with_access(user, :can_view_any_nonconfidential_client_files, :can_view_any_confidential_client_files)
+    enrollment_scope = enrollment_scope.joins(:client).where(client: { id: client_ids }) if client_ids.present?
```

Comment:
Personally I favor this style of using a hash to express a condition on a joined table and it's standard practice in active record.

However we have a practice to using arel instead with global variables (c_t). So I think this is more compliant with our practices:

```suggestion
    enrollment_scope = enrollment_scope.joins(:client).where(c_t[:id].in(client_ids)) if client_ids.present?
```

Reply:
The arel version protects against our non-standard (for rails) table naming conventions.  I prefer the hash style _if_ the table name and relation name match.

Reply:
In this case they do match, plus or minus a capital letter ... so I take it this means we should keep the hash style here?

Reply:
No, the OP code **always uses the arel version**.  The capitalization and pluralization are the main reasons to use it.

Reply:
I see, thanks!